### PR TITLE
[Spark-19726][SQL] Faild to insert null timestamp value to mysql using spark jdbc

### PR DESCRIPTION
**Pull request by shuangshuangwang**
## What changes were proposed in this pull request?

when creating table like following:
> create table timestamp_test(id int(11), time_stamp timestamp not null default current_timestamp);

The result of Excuting "insert into timestamp_test values (111, null)" is different between Spark and JDBC.
```
mysql> select * from timestamp_test;
+------+---------------------+
| id   | time_stamp          |
+------+---------------------+
|  111 | 1970-01-01 00:00:00 | -> spark
|  111 | 2017-06-27 19:32:38 | -> mysql
+------+---------------------+
2 rows in set (0.00 sec)
```
   Because in such case ```StructField.nullable``` is false, so the generated codes of ```InvokeLike``` and ```BoundReference``` don't check whether the field is null or not. Instead, they directly use ```CodegenContext.INPUT_ROW.getLong(1)```, however, ```UnsafeRow.setNullAt(1)``` will put 0 in the underlying memory.
   
   The PR will ```always``` set ```StructField.nullable```  true after obtaining metadata from jdbc connection, Since we can insert null to not null timestamp column in MySQL. In this way, spark will propagate null to underlying DB engine, and let DB to choose how to process NULL.

## How was this patch tested?

Added tests.

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/sources/JdbcInsertSuite.scala [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+      sql(
+        s"""
+           |INSERT INTO timestamp_test values(111, null)
+      """.stripMargin)
+    }.getMessage
+
+    assert(
+      message.contains("NULL not allowed for column \"TIME_STAMP\""),
+      "It is not allowed to insert null into timestamp column which is defined not null."
+    )
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/JdbcInsertSuite.scala]: No need to add a new test suite. Just need to add a test case in `JDBCWriteSuite.scala`

```
  test("INSERT null to a NOT NULL column") {
    val e = intercept[SparkException] {
      sql("INSERT INTO PEOPLE1 values (null, null)")
    }.getMessage
    assert(e.contains("NULL not allowed for column \"NAME\""))
  }
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala [hunk 2]
```diff
@@ -61,7 +61,13 @@ object JDBCRDD extends Logging {
       try {
         val rs = statement.executeQuery()
         try {
-          JdbcUtils.getSchema(rs, dialect)
+          val metaStructType = JdbcUtils.getSchema(rs, dialect)
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala]: Let us add a new boolean parameter `alwaysNullable`  to `getSchema`

**Comment by gatorsmile** [comment]
ok to test

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala (hunk 2 above)
Comment (shuangshuangwang) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala]: Hi  @gatorsmile ，
Thanks for the comments. I will push code changes as you suggested later.

**Comment by shuangshuangwang** [comment]
Hi @gatorsmile ,
I have just pushed code changes.

**Comment by shuangshuangwang** [comment]
Fix scala style tests errors.

**Comment by shuangshuangwang** [comment]
Sorry for scala style checks errors again.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala [hunk 3]
```diff
@@ -273,6 +273,10 @@ object JdbcUtils extends Logging {
     val rsmd = resultSet.getMetaData
     val ncols = rsmd.getColumnCount
     val fields = new Array[StructField](ncols)
+    // if true, spark will propagate null to underlying
+    // DB engine instead of using type default value
+    val alwaysNullable = true
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: put it as the parameter of `getSchema`

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCWriteSuite.scala [hunk 4]
```diff
@@ -20,10 +20,10 @@ package org.apache.spark.sql.jdbc
 import java.sql.{Date, DriverManager, Timestamp}
 import java.util.Properties
 
-import scala.collection.JavaConverters.propertiesAsScalaMapConverter
+import org.apache.spark.SparkException
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCWriteSuite.scala]: Move this to line 27

**Comment by gatorsmile** [comment]
To run the Scala style check, you can run the command in your computer:
> dev/lint-scala

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala [hunk 5]
```diff
@@ -61,7 +61,7 @@ object JDBCRDD extends Logging {
       try {
         val rs = statement.executeQuery()
         try {
-          JdbcUtils.getSchema(rs, dialect)
+          JdbcUtils.getSchema(rs, dialect, true)
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala]: `alwaysNullable = true`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala [hunk 6]
```diff
@@ -264,12 +264,16 @@ object JdbcUtils extends Logging {
   }
 
   /**
-   * Takes a [[ResultSet]] and returns its Catalyst schema.
-   *
-   * @return A [[StructType]] giving the Catalyst schema.
-   * @throws SQLException if the schema contains an unsupported type.
-   */
-  def getSchema(resultSet: ResultSet, dialect: JdbcDialect): StructType = {
+    * Takes a [[ResultSet]] and returns its Catalyst schema.
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: Indents is not right. 
```
/**
 *
 */

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala [hunk 7]
```diff
@@ -264,12 +264,16 @@ object JdbcUtils extends Logging {
...
+    *
+    * @param alwaysNullable - If true, spark will propagate null
+    *                       to underlying DB engine instead of using type default value
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: > @param alwaysNullable If true, all the columns are nullable.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala [hunk 8]
```diff
@@ -264,12 +264,16 @@ object JdbcUtils extends Logging {
...
+    * @return A [[StructType]] giving the Catalyst schema.
+    * @throws SQLException if the schema contains an unsupported type.
+    */
+  def getSchema(resultSet: ResultSet,
+                dialect: JdbcDialect,
+                alwaysNullable: Boolean = false): StructType = {
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: Indents are wrong. See https://github.com/databricks/scala-style-guide

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala [hunk 9]
```diff
@@ -290,7 +294,10 @@ object JdbcUtils extends Logging {
             rsmd.getClass.getName == "org.apache.hive.jdbc.HiveResultSetMetaData" => true
         }
       }
-      val nullable = rsmd.isNullable(i + 1) != ResultSetMetaData.columnNoNulls
+      val nullable = if (alwaysNullable)
+        alwaysNullable
+      else
+        rsmd.isNullable(i + 1) != ResultSetMetaData.columnNoNulls
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: Missing curly braces. See https://github.com/databricks/scala-style-guide#curly-braces

**Comment by shuangshuangwang** [comment]
@gatorsmile ，
Thanks very much for your help. I will pay attention to these details after.

**Comment by shuangshuangwang** [comment]
Hi @gatorsmile ，@srowen，
what is " fails due to an unknown error code, -9." means? Does it mean the patch doesn't pass all tests?

**Comment by shuangshuangwang** [comment]
Jenkins, retest this please.

**Comment by wangyum** [comment]
Jenkins, retest this please

**Comment by wangyum** [comment]
Please setting your username and checking your email:
https://help.github.com/articles/setting-your-username-in-git/#platform-linux
https://help.github.com/articles/setting-your-email-in-git/#platform-linux

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala [hunk 10]
```diff
@@ -290,7 +294,11 @@ object JdbcUtils extends Logging {
             rsmd.getClass.getName == "org.apache.hive.jdbc.HiveResultSetMetaData" => true
         }
       }
-      val nullable = rsmd.isNullable(i + 1) != ResultSetMetaData.columnNoNulls
+      val nullable = if (alwaysNullable) {
+        alwaysNullable
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: Nit: -> `true`. Conceptually, they are different.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala [hunk 11]
```diff
@@ -266,10 +266,14 @@ object JdbcUtils extends Logging {
...
    *
+   * @param alwaysNullable If true, all the columns are nullable.
    * @return A [[StructType]] giving the Catalyst schema.
    * @throws SQLException if the schema contains an unsupported type.
    */
-  def getSchema(resultSet: ResultSet, dialect: JdbcDialect): StructType = {
+  def getSchema(
+      resultSet: ResultSet,
+      dialect: JdbcDialect,
+      alwaysNullable: Boolean = true): StructType = {
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: the default value should be `false`

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: You should change the value to `true` in the caller side.

**Comment by shuangshuangwang** [comment]
Hi @gatorsmile ,
I have just pushed code changes as you suggested. Need i to add this value as a property to class JDBCOptions?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala (hunk 5 above)
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala]: -> `JdbcUtils.getSchema(rs, dialect, alwaysNullable = true)`

**Comment by gatorsmile** [comment]
Above is two comments I have

**Comment by gatorsmile** [comment]
You still miss another comment:
https://github.com/apache/spark/pull/18445/files#r124970022

**Comment by shuangshuangwang** [comment]
Hi @gatorsmile,
I don't understand "Nit: -> true. Conceptually, they are different.", Or what do you mean:
```
val nullable = if (alwaysNullable) {
	true
} else {
	rsmd.isNullable(i + 1) != ResultSetMetaData.columnNoNulls
}
```

**Comment by gatorsmile** [comment]
yes.

**Comment by gatorsmile** [comment]
LGTM

**Comment by gatorsmile** [comment]
Thanks! Merging to master.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala (hunk 2 above)
Comment (oleg-smith) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala]: Isn't that bad to hide useful schema info on `getSchema()` level? 
What if nullable=false is usable for other purposes, like query optimization?

**Pull request closed without merging**